### PR TITLE
chore(gatsby): migrate webpack-hmr-hooks-patch to TypeScript

### PR DESCRIPTION
**Pull request by sasurau4**
## Description
Part of #21995 

I think the patch assume to run on browsers and use `eslint-disable` and `@ts-ignore`. If you have any other ideas, I hope you would pointed it out.

## Related Issues

Part of #21995

### CODE REVIEW

**Comment by sasurau4** [comment]
@blainekasten I investigated the reason why CI failed. It's because the transpiled `webpack-hmr-hooks-patch.js` are bit different and the patch is invalidated.

`webpack-hmr-hooks-patch.js` is transpiled to following.
```
"use strict";

const originalFetch = global.fetch;
delete global.fetch;
module.exports = require(`react-hot-loader/webpack`);
global.fetch = originalFetch;
//# sourceMappingURL=webpack-hmr-hooks-patch.js.map
```

On the other hand, `webpack-hmr-hooks-patch.ts` is transpiled to following.
```
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _webpack = _interopRequireDefault(require("react-hot-loader/webpack"));

exports.default = _webpack.default;

/* eslint-disable @typescript-eslint/ban-ts-ignore */
// @ts-ignore: This file assume run at browsers.
const originalFetch = global.fetch; // @ts-ignore

delete global.fetch;
global.fetch = originalFetch;
//# sourceMappingURL=webpack-hmr-hooks-patch.js.map
```

So, the order of deleting and restoring `global.fetch` is broken.
I think it would be better to move `webpack-hmr-hooks-patch.js` to `cache-dir`

File: packages/gatsby/src/utils/webpack-hmr-hooks-patch.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/* eslint-disable @typescript-eslint/ban-ts-ignore */
+
+// @ts-ignore: This file assume run at browsers.
+const originalFetch = global.fetch
+// @ts-ignore
+delete global.fetch
+
+export { default } from "react-hot-loader/webpack"
```
Comment (blainekasten) [inline packages/gatsby/src/utils/webpack-hmr-hooks-patch.ts]: could we change this to a named export and update the import reference?

Comment (blainekasten) [reply packages/gatsby/src/utils/webpack-hmr-hooks-patch.ts]: might not be possible due to the way webpack works, so if not that's fine

File: packages/gatsby/src/utils/webpack-hmr-hooks-patch.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/* eslint-disable @typescript-eslint/ban-ts-ignore */
+
+// @ts-ignore: This file assume run at browsers.
+const originalFetch = global.fetch
+// @ts-ignore
+delete global.fetch
```
Comment (blainekasten) [inline packages/gatsby/src/utils/webpack-hmr-hooks-patch.ts]: can you try something like this? https://github.com/eslint/eslint/pull/763

Comment (blainekasten) [reply packages/gatsby/src/utils/webpack-hmr-hooks-patch.ts]: not sure if this helps. i think these are fine for this file

File: packages/gatsby/src/utils/webpack-hmr-hooks-patch.ts (hunk 1 above)
Comment (sasurau4) [reply packages/gatsby/src/utils/webpack-hmr-hooks-patch.ts]: I tried named export, but validation of webpack config says `loader must be string`.
So I use default export.

File: packages/gatsby/src/utils/webpack-hmr-hooks-patch.ts (hunk 2 above)
Comment (sasurau4) [reply packages/gatsby/src/utils/webpack-hmr-hooks-patch.ts]: I tried it, but the reasons of CI fail in 81a9274 and d49470a are the same.
The error message `Error: You must provide the URL of lib/mappings.wasm by calling SourceMapConsu  mer.initialize({ 'lib/mappings.wasm': ... }) before using SourceMapConsumer` seems caused by https://github.com/gatsbyjs/gatsby/pull/22280#issuecomment-599228141.

**Comment by blainekasten** [comment]
@sasurau4 thank you so much for your exploration here. I think you are right that this file ultimately should remain as a JS file. So here is what I think we should do.

Let's add a comment to the top of the file that says this file should remain as a JS file, with a reason why and link to this PR. I'd be happy to merge that so we know in the future why this one file stayed a JS file.

**Comment by sasurau4** [comment]
@blainekasten OK, Thanks :+1:

**Pull request merged**